Sources/swift-format/Subcommands/LintFormatOptions.swift

```diff
@@ -26,6 +26,18 @@ struct LintFormatOptions: ParsableArguments {
       """)
   var configuration: String?
 
+  /// A JSON string containing the selected ranges to process.
+  ///
+  /// If not specified, the whole file will be formatted.
+  @Option(
+    name: .long,
+    help: """
+      A JSON string containing an array of {"offset":<Int>, "length":<Int>} pairs specifying \
+      the source code ranges to format.
+      """)
+  var selection: String?
```

Comment:
This might just be for testing, but I think we can improve the user interface a little bit here; how about something like `--lines <start line>:<end line>` for line-based ranges and `--offsets <start offset>:<end offset>` for offset-based ranges? (I don't think "length" is a particularly good way to express this on the command line.) Those flags are a bit closer to what clang-format provides for its command line, and we could accept multiple occurrences of it.

To deal with parsing the `s:e` format, you can use swift-argument-parser's `ExpressibleByArgument` protocol instead of JSON decoding. Maybe the option type is just a pair of ints that implements `ExpressibleByArgument` and then the `validate` method constructs the `Selection` from any `--lines` and `--offsets` that it's given.

Reply:
I like the idea of supporting `--lines <start line>:<end line>` as clang-format does. This seems _much_ easier for specifying ranges at the command line. The one issue there is having to convert to `AbsolutePosition`s, since that's what we have with a `Syntax` struct. But we could do that conversion at the start. (I'm already worried about performance of all the range comparisons that happen. We don't want to make that harder.) Anyhow, maybe this can be a second PR?

I was modeling the offset/length bit (kinda) off what clang-format does. It can take multiple `--offset`/`--length` pairs. (Would swift-argument-parser support that? I assumed not, but I admit that I never looked.) I ended up using JSON decoding because it was easy. 🤷‍♂️ clang-format doesn't support `--offsets <start offset>:<end offset>`, so there's no precedent there. My expectation is that a human wouldn't ever specify ranges with offsets (or offsets/lengths). It's just not convenient. But it is convenient for an editor shelling out to swift-format.

If I do manage to change `Selection.Range` to be a `typealias` to `Range<AbsolutePosition>`, then it might be a bit more convenient to pass two offsets, rather than offset/length. Do you have a strong opinion as to whether I should bother with this refactoring? I made a start down that road, and it was a kinda annoying set of changes. 🤣

Reply:
My own personal opinion is that ranges are a bit more natural than an offset/length pair. I'm not sure what motivated clang-format's `--offset` and `--length` flags unless it was some very specific workflow that already had the information in that form, but I don't think ordered pairs of `--offset X --length Y ...` would work well in swift-argument-parser or for human users.

Reply:
FWIW, I would also prefer `--lines <start line>:<end line>`. And if we do the conversion up front with a `SourceLocationConverter`, I don’t think it should be too expensive. `SourceLocationConverter` needs to build up a line table but I would be surprised if that’s the same order of magnitude as formatting (would be nice to confirm though).

Reply:
The `SourceLocationConverter` is already being created in other code paths anyway, so using it to map the line numbers to offsets or vice versa shouldn't create additional overhead; we just need to make sure we're using the same one and not creating a separate one just for that purpose.

Reply:
Yeah, I like `--lines <start line>:<end line>` as well. For an actual user, this is way more useful than offsets. But offsets works well when integrating with other tools. (It lets you, for instance, reformat spacing within a single line without affecting anything else. That's something you might want to do from within an editor.) Chatting with Tony on Friday we also agreed that using `--offsets <start offset>:<end offset>` is cleaner. I'll make that change in this PR. Supporting `--lines <start line>:<end line>` can come in a followup PR.

Reply:
@DaveEwing Alex pointed me to this conversion because I'm trying to implement `git swift-format`, we do need `--lines` for that to work because that's the only easily extractable information from `git diff-index`.